File: src/leetcode/python/medium/ImplementTrie_208.py
class TrieNode(object):
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.is_word = False
        self.children = {}


class ImplementTrie_208(object):
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        """
        Inserts a word into the trie.
        :type word: str
        :rtype: void
        """
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]
        node.is_word = True

    def search(self, word):
        """
        Returns if the word is in the trie.
        :type word: str
        :rtype: bool
        """
        return self.find(word, True)

    def startsWith(self, prefix):
        """
        Returns if there is any word in the trie
        that starts with the given prefix.
        :type prefix: str
        :rtype: bool
        """
        return self.find(prefix, False)

    def find(self, word, whole):
        node = self.root
        for c in word:
            if c not in node.children:
                return False
            node = node.children[c]
        return node.is_word if whole else True

File: src/leetcode/java/ImplementTrie_208.java
package leetcode.java.medium;

public class ImplementTrie_208 {
  private TrieNode root;

  public ImplementTrie_208() {
    root = new TrieNode();
  }

  // Inserts a word into the trie.
  public void insert(String word) {
    char[] arr = word.toCharArray();
    int childIndex;
    TrieNode pt = root;
    for (int idx = 0; idx < arr.length; idx++) {
      childIndex = arr[idx] - 'a';
      if (pt.children[childIndex] == null) {
        pt.children[childIndex] = new TrieNode();
      }
      pt = pt.children[childIndex];
    }
    pt.isWord = true;
  }

  // Returns if the word is in the trie.
  public boolean search(String word) {
    return search(word, true);
  }

  // Returns if there is any word in the trie
  // that starts with the given prefix.
  public boolean startsWith(String prefix) {
    return search(prefix, false);
  }

  private boolean search(String word, boolean whole) {
    char[] arr = word.toCharArray();
    int childIndex;
    TrieNode pt = root;
    for (int idx = 0; idx < arr.length; idx++) {
      childIndex = arr[idx] - 'a';
      if (pt.children[childIndex] == null) return false;
      pt = pt.children[childIndex];
    }
    return whole ? pt.isWord : true;
  }

  class TrieNode {
    TrieNode[] children;
    boolean isWord;

    // Initialize your data structure here.
    public TrieNode() {
      this.children = new TrieNode[26];
      this.isWord = false;
    }
  }
}
